docs/operations.md

```diff
@@ -1,56 +1,82 @@
-# Operations
+# Operations Guide
 
-## Configuring an `EtcdCluster`
+This guide provides instructions aimed at Kubernetes cluster administrators who wish to manage etcd clusters. It
+does not cover initial setup or installation of the operator, and instead assumes that it is already present and working
+correctly.
 
-There is a configuration file for a 3-node Etcd cluster at `config/samples/etcd_v1alpha1_etcdcluster.yaml`.
-In this example each Etcd node has 50Mi storage and uses a [Storage Class](https://kubernetes.io/docs/concepts/storage/storage-classes/#the-storageclass-resource) called "standard".
+The etcd cluster controller uses the Kubernetes API, via
+[Custom Resources](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/), to drive
+administration of etcd clusters. You can administer your cluster manually using `kubectl`, or through any other
+mechanism capable of modifying and viewing Kubernetes resources.
 
-**NOTE**: The `etcd-cluster-operator` can not currently reconcile changes to the storage settings.
-Do not make changes to this field after you have applied the manifest.
-In future this field will be made immutable. See https://github.com/improbable-eng/etcd-cluster-operator/issues/49
+## Creating a new Cluster
 
-## Examples
+To create a new cluster, create an `EtcdCluster` resource in the namespace you want the cluster's pods to run in. The
+`spec` field of the resource is used to configure the desired properties of the cluster.
 
-Here are some examples of day-to-day operations and explanations of how Etcd data is handled in each case.
+### Replicas
 
-### Bootstrap a New Cluster
+The `spec.replicas` field determines the number of pods that are run in the etcd cluster. It is [strongly suggested that
+this be an odd number](https://etcd.io/docs/v3.4.0/faq/#why-an-odd-number-of-cluster-members). In etcd [adding more
+replicas will degrade write performance](https://etcd.io/docs/v3.4.0/faq/#what-is-maximum-cluster-size). This can be `1`
+for a testing environment, but for durability it is suggested that this is at least `3`. In most situations `5` is the
+highest sensible setting, although neither etcd nor this operator impose a limit.
```

Comment:
> this seems to be optimising for availability […]

Yes. Running an even number of replicas is strictly worse for durability than running the equivalent n-1 odd numbered cluster.

> Although adding a node to an odd-sized cluster appears better since there are more machines, the fault tolerance is worse since exactly the same number of nodes may fail without losing quorum but there are more nodes that can fail.

(From the [etcd FAQ](https://etcd.io/docs/v3.4.0/faq/#why-an-odd-number-of-cluster-members))

So any reason to run an even number of peers would have to be a good one with that limitation in mind.

> ...people might want to run one peer per k8s node, in which case an [even] number of replicas could be fine

The data locality case is interesting, but I don't think the operator really supports this use-case for a number of reasons:

* We we're planning to add the ability to pass through taint/affinity information for the pods, but effectively replicating the behaviour of a `DemonSet` to put one pod on each node isn't something we've considered.
* This wouldn't work beyond about nine or so nodes. So if you had a cluster with lots of nodes your etcd cluster is very quickly going to run into serious troubles as it gets too big.
* Even if we did get one etcd pod on each node, the current headless service approach just gives you a list of pod IPs. A client wouldn't be able to tell which one is on the same node as it. So it might just dial an etcd pod on a different node anyway.
* Depending on the exact application, I wonder if a separate caching layer would be more appropriate. Deployed as a sidecar or even a demonset for node locality.

I could add some of this discussion to the operations guide, perhaps just pointing out that these deployment instructions are optimised for availability following the etcd documentation.

Is the data locality use-case one you're concerned about in your application? We could have a deeper discussion if you'd like to address this use-case in particular.

Reply:
> pointing out that these deployment instructions are optimised for availability following the etcd documentation.

yep i think that's fine

> Is the data locality use-case one you're concerned about in your application?

no it's not something we currently need - was just a thought

Reply:
> yep i think that's fine

Will do then.

> no it's not something we currently need - was just a thought

It's not a bad idea, I just think we'd need to do more work than that to support that use-case. It could be done if you needed it.